lib/katello/tasks/clean_installed_packages.rake

```diff
@@ -0,0 +1,52 @@
+namespace :katello do
+  desc "Clean duplicate installed packages"
+  task :clean_installed_packages => :environment do
+    User.current = User.anonymous_admin
+
+    puts "Populating cache"
+    cache = PackageCache.new
+    cache.populate
+
+    total = ::Host.count
+    count = 1
+    ::Host.find_each do |host|
+      puts "Host #{count}/#{total}"
+
+      if host.installed_package_ids.any?
+        nvras = host.installed_packages.pluck(:nvra)
+        ids = nvras.map { |nvra| cache.fetch(nvra) }
+        existing_ids = host.installed_package_ids.to_a
+        unless existing_ids.sort == ids.sort
+          host.sync_package_associations(ids)
+        end
+      end
+      count += 1
+    end
```

Comment:
Its determining the nvras for all the installed packages for that host, then finding all the installed package ids from our cache that have those nvras (this is the single set that will be kept and not deleted.    Finally it calls the sync_package_associations to change the associations to the InstalledPackage ids we want to keep.  Does that answer your question?